heron/stmgr/src/cpp/manager/tmaster-client.cpp

```diff
@@ -217,5 +235,48 @@ void TMasterClient::SendHeartbeatRequest() {
   return;
 }
 
+void TMasterClient::SavedInstanceState(const proto::system::Instance& _instance,
+                                       const std::string& _checkpoint_id) {
+  proto::ckptmgr::InstanceStateStored message;
+  message.set_checkpoint_id(_checkpoint_id);
+  message.mutable_instance()->CopyFrom(_instance);
+  SendMessage(message);
+}
+
+void TMasterClient::SendRestoreTopologyStateResponse(proto::system::StatusCode _status,
+                                                     const std::string& _ckpt_id,
+                                                     sp_int64 _txid) {
+  proto::ckptmgr::RestoreTopologyStateResponse message;
+  message.mutable_status()->set_status(_status);
+  message.set_checkpoint_id(_ckpt_id);
+  message.set_restore_txid(_txid);
+  SendMessage(message);
+}
+
+void TMasterClient::HandleRestoreTopologyStateRequest(
+              proto::ckptmgr::RestoreTopologyStateRequest* _message) {
+  restore_topology_watch_(_message->checkpoint_id(), _message->restore_txid());
+  __global_protobuf_pool_release__(_message);
+}
+
+void TMasterClient::HandleStartStmgrStatefulProcessing(
```

Comment:
So topologies not doing exactly once can start processing right away as soon as assignment are propagated to all stmgrs/instances. However for exactly once semantics, we have to wait until all of the stmgrs/instances are restored to a certain globally consistent checkpoint before proceeding. Once that determination is done by tmaster, this is the message that it sends to the stmgr to start the actual processing